### PR TITLE
Issue 34817 db exception IsTransient property

### PR DESCRIPTION
**Pull request by rgarrison12345**
Adding boolean property to DbException with a default implementation of false.  This property determines if the previously executed command is transient and could succeed if executed again without any changes.

### CODE REVIEW

**Comment by Wraith2** [comment]
You need to change the ref project to have the same change or it won't be exposed.
I don't think you need to explicitly set the value of the property because false is the default.

**Comment by danmosemsft** [comment]
@rgarrison12345 we generally don't put the issue # in the title. In the body (the top block of text) you can put `Fixes #34817` or `Relates to #34817` depending on whether merging this PR should close that issue or not.

**Comment by danmosemsft** [comment]
@rgarrison12345 sorry, but I just noticed that #34817 is marked "api-ready-for-review". Before we have a PR, it needs to be approved as a new API, then it will be labled "api-approved". We would welcome a PR at that point (possibly you could reopen this one). Thanks for contributing.

**Comment by rgarrison12345** [comment]
Hello @danmosemsft,

I opened a pull request before the API had been agreed upon because I had asked @roji  if it was ok to do so in the thread for the original issue because the change was small and it's a non-breaking change.  His thought was that because the change was small despite not being marked ready for review I could go ahead and proceed.

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs [hunk 1]
```diff
@@ -2170,6 +2170,10 @@ public partial class DbEnumerator : System.Collections.IEnumerator
     }
     public abstract partial class DbException : System.Runtime.InteropServices.ExternalException
     {
+        /// <summary>
+        /// When overriden in a derived class return true in all cases where a simple retry of the operation without any other change may be successful.
+        /// </summary>
+        public virtual bool IsTransient { get; protected set; }
```
Comment (roji) [inline src/libraries/System.Data.Common/ref/System.Data.Common.cs:2176]: Please define this property like other properties in the ref file:

```c#
public virtual bool IsTransient { get { throw null; } }
```

This isn't an implementation, just a contract - so it only needs to compile. Also, please move this down under the constructors as is the convention with other types here.

File: src/libraries/System.Data.Common/ref/System.Data.Common.cs [hunk 2]
```diff
@@ -2170,6 +2170,10 @@ public partial class DbEnumerator : System.Collections.IEnumerator
     }
     public abstract partial class DbException : System.Runtime.InteropServices.ExternalException
     {
+        /// <summary>
```
Comment (roji) [inline src/libraries/System.Data.Common/ref/System.Data.Common.cs:2173]: Right now, documentation is added separately in https://github.com/dotnet/dotnet-api-docs and not here - that's why nothing currently has XML documentation. So let's remove this for now (from ref as well). Once this PR is merged we can do a separate PR to that repo, and if one day documentation is added here, we can do it for all the System.Data docs together.

File: src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs [hunk 3]
```diff
@@ -8,6 +8,11 @@ namespace System.Data.Common
     [System.Runtime.CompilerServices.TypeForwardedFrom("System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public abstract class DbException : System.Runtime.InteropServices.ExternalException
     {
+        /// <summary>
+        /// When overriden in a derived class return true in all cases where a simple retry of the operation without any other change may be successful.
+        /// </summary>
+        public virtual bool IsTransient { get; protected set; }
```
Comment (roji) [inline src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs:14]: This shouldn't be an auto-property, but rather a simple getter-only property. This allows implementations greater freedom in how they implement this, and doesn't add a backing field which usually won't be needed. So we need something like:

```c#
public virtual bool IsTransient => false;
```

**Review by roji — COMMENTED** [review]
@rgarrison12345 here are some comments. Don't worry about this PR being closed for now, we can reopen it once the API is approved (or resubmit). Though it's true that if you don't want the risk of doing work and the issue not getting approval, it may be better to wait.

File: src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs (hunk 3 above)
Comment (rgarrison12345) [reply src/libraries/System.Data.Common/src/System/Data/Common/DbException.cs:14]: Thanks @roji when the API is approved I will make these changes.
